File: benchmark_core/python/bench_mark.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#
import os
from time import perf_counter
from dataclasses import dataclass
import json

from tqdm import tqdm
import psutil
import csv
from tsfile import TSDataType, ColumnCategory
from tsfile import TableSchema, ColumnSchema
from tsfile import Tablet
from tsfile import TsFileTableWriter, TsFileReader

@dataclass
class Config:
    tablet_num: int
    tag1_num: int
    tag2_num: int
    timestamp_per_tag: int
    field_type_vector: list[int]

def load_config(config_path: str) -> Config:
    with open(config_path, 'r') as f:
        config_data = json.load(f)
    return Config(**config_data)

type_list = [TSDataType.INT32, TSDataType.INT64, TSDataType.FLOAT, TSDataType.DOUBLE, TSDataType.BOOLEAN]


def print_config(config: Config):
    data_types_name = ["INT64", "INT32", "FLOAT", "DOUBLE", "BOOLEAN"]
    print("=====================")
    print("TsFile benchmark For Python")
    print("Schema Configuration:")
    print(f"Tag Column num: {2}")
    print(f"TAG1 num: {config.tag1_num} TAG2 num: {config.tag2_num}\n")

    print("Filed Column and types: ")
    column_num = 0
    for i in range(5):
        print(f"{data_types_name[i]}x{config.field_type_vector[i]}  ", end="")
        column_num += config.field_type_vector[i]

    print("\n")
    print(f"Tablet num: {config.tablet_num}")
    print(f"Tablet row num per tag: {config.timestamp_per_tag}")

    total_points = (config.tablet_num *
                    config.tag1_num *
                    config.tag2_num *
                    config.timestamp_per_tag *
                    column_num)
    print(f"Total points is {total_points}")
    print("======")

column_name = []

def get_memory_usage_kb():
    process = psutil.Process(os.getpid())
    return process.memory_info().rss // 1024

def bench_mark_write():
    csv_file = "memory_usage_python.csv"

    csv_writer = csv.writer(open(csv_file, "w"))
    csv_writer.writerow(["iter_num", "memory_usage(kb)"])
    iter_num = 0
    print_config()
    column_schema_list = []
    column_datat_type = []
    column_schema_list.append(ColumnSchema("TAG1", TSDataType.STRING, ColumnCategory.TAG))
    column_name.append("TAG1")
    column_datat_type.append(TSDataType.STRING)
    column_schema_list.append(ColumnSchema("TAG2", TSDataType.STRING, ColumnCategory.TAG))
    column_name.append("TAG2")
    column_datat_type.append(TSDataType.STRING)


    i = 2
    for count, type in zip(config.field_type_vector, type_list):
        for _ in range(count):
            column_schema_list.append(ColumnSchema("FIELD" + str(i), type, ColumnCategory.FIELD))
            column_name.append("FIELD" + str(i))
            column_datat_type.append(type)
            i = i + 1

    timestamp = 0
    table_schema = TableSchema("TestTable", column_schema_list)
    start = perf_counter()
    prepare_time = 0
    writing_time = 0
    csv_writer.writerow([iter_num, get_memory_usage_kb()])
    iter_num += 1
    with TsFileTableWriter("tsfile_table_write_bench_mark.tsfile", table_schema) as writer:
        csv_writer.writerow([iter_num, get_memory_usage_kb()])
        iter_num += 1
        for i in tqdm(range(config.tablet_num), desc="Tablets"):
            csv_writer.writerow([iter_num, get_memory_usage_kb()])
            iter_num += 1
            row_num = 0
            prepare_start = perf_counter()
            tablet = Tablet(column_name, column_datat_type,
                            config.timestamp_per_tag * config.tag1_num *
                            config.tag2_num)

            for j in range(config.tag1_num):
                for k in range(config.tag2_num):
                    for row in range(config.timestamp_per_tag):
                        tablet.add_timestamp(row_num, timestamp + row)
                        tablet.add_value_by_index(0, row_num, "tag1_" + str(j))
                        tablet.add_value_by_index(1, row_num, "tag2_" + str(k))
                        for col in range(2, len(column_name)):
                            if column_datat_type[col] == TSDataType.INT32:
                                tablet.add_value_by_index(col, row_num, timestamp)
                            elif column_datat_type[col] == TSDataType.INT64:
                                tablet.add_value_by_index(col, row_num, timestamp)
                            elif column_datat_type[col] == TSDataType.FLOAT:
                                tablet.add_value_by_index(col, row_num, timestamp * 1.1)
                            elif column_datat_type[col] == TSDataType.DOUBLE:
                                tablet.add_value_by_index(col, row_num, timestamp * 1.1)
                            elif column_datat_type[col] == TSDataType.BOOLEAN:
                                tablet.add_value_by_index(col, row_num, timestamp % 2 == 0)
                        row_num = row_num + 1

            prepare_time += perf_counter() - prepare_start
            write_start = perf_counter()
            csv_writer.writerow([iter_num, get_memory_usage_kb()])
            iter_num += 1
            writer.write_table(tablet)
            writing_time += perf_counter() - write_start
            timestamp = timestamp + config.timestamp_per_tag
            csv_writer.writerow([iter_num, get_memory_usage_kb()])
            iter_num += 1

    csv_writer.writerow([iter_num, get_memory_usage_kb()])
    iter_num += 1
    end = perf_counter()
    total_time = end - start
    size = os.path.getsize("tsfile_table_write_bench_mark.tsfile")

    total_points = config.tablet_num * config.tag1_num * config.tag2_num * \
                   config.timestamp_per_tag * len(column_name)

    print("finish bench mark for python")
    print(f"tsfile size is {size} bytes ~ {size // 1024}KB")

    print(f"prepare data time is {prepare_time:.6f} s")
    print(f"writing data time is {writing_time:.6f} s")
    print(f" total_time is {total_time} s")
    writing_speed = int(total_points / (prepare_time + writing_time))
    print(f"writing speed is {writing_speed} points/s")

    total_time_seconds = (end - start)
    print(f"total time is {total_time_seconds:.6f} s")

def bench_mark_read():
    start = perf_counter()
    row = 0
    with TsFileReader("tsfile_table_write_bench_mark.tsfile") as reader:
        result = reader.query_table("TestTable", column_name)
        first = True
        while result.next():
            row = row + 1

    end = perf_counter()
    total_time = end - start
    reading_speed = int(row * len(column_name) / total_time)
    print("total row is ", row)
    print(f"reading data time is {total_time} s")
    print(f"reading data speed is {reading_speed} points/s")


bench_mark_write()
bench_mark_read()

File: benchmark_core/java/src/main/java/org/apache/tsfile/benchmark/BenchMark.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.tsfile.benchmark;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.apache.tsfile.enums.TSDataType;
import org.apache.tsfile.exception.read.ReadProcessException;
import org.apache.tsfile.exception.write.NoMeasurementException;
import org.apache.tsfile.exception.write.NoTableException;
import org.apache.tsfile.exception.write.WriteProcessException;
import org.apache.tsfile.file.metadata.ColumnSchema;
import org.apache.tsfile.file.metadata.ColumnSchemaBuilder;
import org.apache.tsfile.file.metadata.TableSchema;
import org.apache.tsfile.fileSystem.FSFactoryProducer;
import org.apache.tsfile.read.query.dataset.ResultSet;
import org.apache.tsfile.read.v4.ITsFileReader;
import org.apache.tsfile.read.v4.TsFileReaderBuilder;
import org.apache.tsfile.write.record.Tablet;
import org.apache.tsfile.write.v4.ITsFileWriter;
import org.apache.tsfile.write.v4.TsFileWriterBuilder;

import org.apache.tsfile.benchmark.ConfigLoad;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class BenchMark {
    private static final Logger LOGGER = LoggerFactory.getLogger(BenchMark.class);

    public static void main(String[] args) throws Exception {
        ConfigLoad configLoad = ConfigLoad.Load("tmp/json");
        System.out.println(configLoad);

        MemoryMonitor monitor = new MemoryMonitor();
        String path = "tsfile_table_write_bench_mark_java.tsfile";
        File f = FSFactoryProducer.getFSFactory().getFile(path);
        if (f.exists()) {
            Files.delete(f.toPath());
        }
        monitor.recordMemoryUsage();
        List<String> column_names = new ArrayList<>();
        List<TSDataType> column_types = new ArrayList<>();
        List<ColumnSchema> columnSchemas = new ArrayList<>();
        columnSchemas.add(new ColumnSchemaBuilder().name("TAG1").dataType(TSDataType.STRING).category(Tablet.ColumnCategory.TAG).build());
        columnSchemas.add(new ColumnSchemaBuilder().name("TAG2").dataType(TSDataType.STRING).category(Tablet.ColumnCategory.TAG).build());
        column_names.add("TAG1");
        column_names.add("TAG2");
        column_types.add(TSDataType.STRING);
        column_types.add(TSDataType.STRING);

        int fieldIndex = 2;
        for (int i = 0; i < configLoad.field_type_vector.size(); i++) {
            int count = configLoad.field_type_vector.get(i);
            TSDataType dataType = configLoad.getTSDataType(i);
            for (int j = 0; j < count; j++) {
                columnSchemas.add(new ColumnSchemaBuilder().name("FIELD" + fieldIndex).dataType(dataType).category(Tablet.ColumnCategory.FIELD).build());
                column_names.add("FIELD" + fieldIndex);
                column_types.add(dataType);
                fieldIndex++;
            }
        }
        monitor.recordMemoryUsage();
        long totalPrepareTimeNs = 0;
        long totalWriteTimeNs = 0;
        long writeStartTime = 0;
        TableSchema tableSchema = new TableSchema("TestTable", columnSchemas);
        monitor.recordMemoryUsage();
        try (ITsFileWriter writer = new TsFileWriterBuilder().file(f).tableSchema(tableSchema).build()) {
            monitor.recordMemoryUsage();
            long timestamp = 0;
            for (int table_ind = 0; table_ind < configLoad.tablet_num; table_ind++) {
                long prepareStartTime = System.nanoTime();
                Tablet tablet = new Tablet(column_names, column_types, configLoad.tag1_num * configLoad.tag2_num * configLoad.timestamp_per_tag);
                int row_count = 0;
                for (int tag1_ind = 0; tag1_ind < configLoad.tag1_num; tag1_ind++) {
                    for (int tag2_ind = 0; tag2_ind < configLoad.tag2_num; tag2_ind++) {
                        for (int row = 0; row < configLoad.timestamp_per_tag; row++) {
                            tablet.addTimestamp(row_count, timestamp + row);
                            tablet.addValue(row_count, 0, "tag1_" + tag1_ind);
                            tablet.addValue(row_count, 1, "tag2_" + tag2_ind);

                            for (int i = 2; i < column_types.size(); i++) {
                                switch (column_types.get(i)) {
                                    case INT32:
                                        tablet.addValue(row_count, i, (int) timestamp);
                                        break;
                                    case INT64:
                                        tablet.addValue(row_count, i, timestamp);
                                        break;
                                    case FLOAT:
                                        tablet.addValue(row_count, i, (float) (timestamp * 1.1));
                                        break;
                                    case DOUBLE:
                                        tablet.addValue(row_count, i, (double) timestamp * 1.1);
                                        break;
                                    case BOOLEAN:
                                        tablet.addValue(row_count, i, timestamp % 2 == 0);
                                    default:
                                        //
                                }
                            }
                            row_count++;
                        }
                    }
                }
                monitor.recordMemoryUsage();
                long prepareEndTime = System.nanoTime();

                totalPrepareTimeNs += (prepareEndTime - prepareStartTime);
                writeStartTime = System.nanoTime();
                writer.write(tablet);
                monitor.recordMemoryUsage();
                long writeEndTime = System.nanoTime();
                totalWriteTimeNs += (writeEndTime - writeStartTime);
                timestamp += configLoad.timestamp_per_tag;
            }
            writeStartTime = System.nanoTime();
        } catch (WriteProcessException e) {
            LOGGER.error("meet error in TsFileWrite ", e);
        }
        long writeEndTime = System.nanoTime();
        totalWriteTimeNs += (writeEndTime - writeStartTime);
        monitor.recordMemoryUsage();
        long end = System.nanoTime();
        double totalPrepareTimeSec = totalPrepareTimeNs / 1_000_000_000.0;
        double totalWriteTimeSec = totalWriteTimeNs / 1_000_000_000.0;
        long size = f.length();

        monitor.close();
        System.out.println("Finish bench mark for java");
        System.out.printf("Tsfile size is %d bytes ~ %dKB%n", size, size / 1024);

        System.out.printf("Prepare data time is %.6f s%n", totalPrepareTimeSec);
        System.out.printf("Writing data time is %.6f s%n", totalWriteTimeSec);

        long totalPoints = (long) configLoad.tablet_num * configLoad.tag1_num * configLoad.tag2_num * configLoad.timestamp_per_tag * (column_names.size() - 2);
        double writingSpeed = totalPoints / (totalPrepareTimeSec + totalWriteTimeSec);
        System.out.printf("writing speed is %d points/s%n", (long) writingSpeed);
        Map<String, Object> result = new HashMap<>();
        result.put("tsfile_size", size / 1024);
        result.put("prepare_time", totalPrepareTimeSec);
        result.put("write_time", totalWriteTimeSec);
        result.put("writing_speed", writingSpeed);


        Integer row = 0;
        long read_start = System.nanoTime();
        try (ITsFileReader reader = new TsFileReaderBuilder().file(f).build()) {
            ResultSet resultSet = reader.query("TestTable", column_names, Long.MIN_VALUE, Long.MAX_VALUE);
            while (resultSet.next()) {
                row++;
            }
        }
        System.out.println("row is " + row);
        long read_end = System.nanoTime();
        double totalReadTimeSec = (read_end - read_start) / 1_000_000_000.0;
        System.out.printf("read time is %.6f s%n", totalReadTimeSec);
        Long readSpeed = Math.round(row * column_names.size() / totalReadTimeSec);
        System.out.printf("read speed is %.6f points/s %n", readSpeed);
        result.put("reading_time", totalReadTimeSec);
        result.put("reading_speed", readSpeed);
        ObjectMapper mapper = new ObjectMapper();
        mapper.writerWithDefaultPrettyPrinter().writeValue(new File("/result/results_java.json"), result);
    }
}
